File: python/230-Kth Smallest Element in a BST.py
from bintree_helper import *

'''
For followup:
Use B+ Tree.
'''

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        "Recursive Inorder-Traversal, Time: O(H+k), Space: O(H)"
        count = 0
        result = 0
        def dfs(node):
            nonlocal count, result
            if node.left: dfs(node.left)
            if count >= k: return
            count += 1
            result = node.val
            if node.right: dfs(node.right)
        dfs(root)
        return result
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        "Iterative Inorder-Traversal, Time: O(H+k), Space: O(H)"
        def helper(node, stack):
            while node:
                stack.append(node)
                node = node.left
        from collections import deque
        stack = deque()
        helper(root, stack)
        result = 0
        while stack and k:
            cur = stack.pop()
            result = cur.val
            k -= 1
            if cur.right: helper(cur.right, stack)
        return result

root = deserialize("[3,1,4,null,2]")
k = 1
result = Solution().kthSmallest(root, k)
assert result == 1

root = deserialize("[5,3,6,2,4,null,null,1]")
k = 3
result = Solution().kthSmallest(root, k)
assert result == 3

File: cpp/230. Kth Smallest Element in a BST.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// Recursive inorder traversal, Time: O(n), Space: O(n)
public:
    void dfs(TreeNode *root, vi &arr) {
        if(!root) return;
        dfs(root->left, arr);
        arr.push_back(root->val);
        dfs(root->right, arr);
    }
    int kthSmallest(TreeNode* root, int k) {
        vi arr;
        dfs(root, arr);
        return arr[k-1];
    }
};

class Solution {
// Iterative inorder traversal, Time: O(k), Space: O(n)
public:
    void helper(TreeNode *node, stack<TreeNode*> &s) {
        while(node) {
            s.push(node);
            node = node->left;
        }
    }
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode*> s;
        helper(root, s);

        int ans = 0;
        while(k--) {
            TreeNode *node = s.top(); s.pop();
            ans = node->val;
            if(node->right) helper(node->right, s);
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

}